idseq_pipeline/commands/assembly_functions.py

```diff
@@ -0,0 +1,134 @@
+    # download input
+    execute_command("aws s3 cp --quiet %s/%s %s/" % (ALIGNMENT_S3_PATH, TAXON_COUNTS, INPUT_DIR))
+    pipeline_output_json = os.path.join(INPUT_DIR, TAXON_COUNTS)
+    execute_command("aws s3 cp --quiet %s/%s %s/" % (POSTPROCESS_S3_PATH, TAXID_ANNOT_FASTA, INPUT_DIR))
+    full_fasta = os.path.join(INPUT_DIR, TAXID_ANNOT_FASTA)
+
+    # run assembly
+    def get_taxids_to_assemble(taxon_counts):
+        total_counts = {} # sum of NT and NR counts for each taxid
+        filtered_total_counts = {} # taxids that satisfy count threshold
+        for item in taxon_counts:
+            taxid = item['tax_id']
+            total_counts[taxid] = total_counts.get(taxid, 0) + item['count']
+            if total_counts[taxid] >= ASSEMBLY_READ_THRESHOLD:
+                filtered_total_counts[taxid] = total_counts[taxid]
+        sorted_filtered_taxids = sorted(filtered_total_counts, key=filtered_total_counts.get, reverse=True)        
+        return sorted_filtered_taxids
+
+    def make_inputs_for_assembly():
+        # Get taxids to assemble based on criteria from report, currently: taxids with read count >= ASSEMBLY_READ_THRESHOLD
+        with open(pipeline_output_json) as f:
+            pipeline_output = json.load(f)
+        taxon_counts = pipeline_output['pipeline_output']['taxon_counts_attributes']
+        sorted_taxids_to_assemble = get_taxids_to_assemble(taxon_counts)
+        # Get reads for those taxids
+        output = {}
+        hit_delimiters = ['_nt', '_nr'] # annotations are like e.g. "genus_nt:543:"
+        for taxid in sorted_taxids_to_assemble:
+            pattern = '\|'.join(['%s:%s:' % (delimiter, taxid) for delimiter in hit_delimiters])
+            partial_fasta =  os.path.join(RESULT_DIR, taxid + ".fasta")
+            try:
+                subprocess.check_call("grep -A 1 --no-group-separator '%s' %s > %s" % (pattern, full_fasta, partial_fasta), shell=True)
+                output[taxid] = partial_fasta
+            except:
+                print "WARNING: taxid %s was not found in the annotated fasta" % taxid
+        # Also include the full fasta as an input to assembly
+        output['all'] = full_fasta
+        sorted_taxids_to_assemble.append('all')
+        sorted_taxids_to_assemble = [item for item in sorted_taxids_to_assemble if item in output.keys()]
+        return output, sorted_taxids_to_assemble
+
+    def length_without_newlines(sequence):
+        return len(sequence.replace("\n",""))
+
+    def read_fasta(input_fasta):
+        ''' Read fasta file into dictionary. Newlines remain part of the keys/values. '''
+        read_name = None
+        d = {}
+        read_order = []
+        with open(input_fasta, 'rb') as f:
+            for line in f:
+                if line.startswith('>'):
+                    read_name = line
+                    read_order += [read_name]
+                else:
+                    d[read_name] = d.get(read_name, '') + line
+        return d, read_order
```

Comment:
Some fasta files are 30GB uncompressed.   How large are the files this is meant to work for?

Reply:
This is used only on the annotated non-host fasta, which contains at most 2M reads (due to subsampling), corresponding to about 650 MB in practice. Still, I will see if I can get rid of read_order somehow

Reply:
Ok, I rewrote it so that it doesn't read the file into memory